File: src/main/java/io/github/ytg1234/fabman/dataspec/FabmanPackage.java
package io.github.ytg1234.fabman.dataspec;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class FabmanPackage {
	private String name;
	private String mavenUrl;
	private String group;
	private String artifact;
	private List<String> configurations;

	public FabmanPackage(@Nullable String name, String mavenUrl, String group, String artifact, List<String> configurations) {
		this.name = name;
		this.mavenUrl = mavenUrl;
		this.group = group;
		this.artifact = artifact;
		this.configurations = configurations;
	}

	@NotNull
	public String getGroup() {
		return group;
	}

	@NotNull
	public String getArtifact() {
		return artifact;
	}

	@Nullable
	public String getName() {
		return name;
	}

	@NotNull
	public String getMavenUrl() {
		return mavenUrl;
	}

	@Override
	public String toString() {
		return "Package slug is " + name + ", Gradle ID is " + group + ":" + artifact + ":<version>, and required maven repo is " + mavenUrl;
	}

	public List<String> getConfigurations() {
		return configurations;
	}
}

File: src/main/kotlin/io/github/ytg1234/fabman/dataspec/FabmanPackage.kt
package io.github.ytg1234.fabman.dataspec

import kotlinx.serialization.Serializable

@Serializable
data class FabmanPackageImpl(
    override val name: String = "",
    override val mavenUrl: String = "",
    override val group: String,
    override val artifact: String,
    override val configurations: List<String>
) : FabmanPackage

interface FabmanPackage {
    val name: String?
    val mavenUrl: String
    val group: String
    val artifact: String
    val configurations: List<String>

    companion object {
        operator fun invoke(
            name: String = "",
            mavenUrl: String = "",
            group: String,
            artifact: String,
            configurations: List<String>
        ) = FabmanPackageImpl(name, mavenUrl, group, artifact, configurations)

        fun serializer() = FabmanPackageImpl.serializer()
    }
}
